src/build.d

```diff
@@ -0,0 +1,1131 @@
+    );
+    if (res.helpWanted)
+    {
+        defaultGetoptPrinter(`./build.d <targets>...
+
+Examples:
+
+    ./build.d dmd                                                  # build DMD
+    ./build.d unittest                                             # runs internal unittests
+    ./build.d clean                                                # remove all generated files
+
+Options:
+`, res.options);
+        "\nSee the README.md for a more in-depth explanation of the build script.".writeln;
+        return;
+    }
+
+    // parse arguments
+    args.popFront;
+    args2Environment(args);
+
+    // bootstrap all needed environment variables
+    env = getEnvironment;
+
+    // get all sources
+    sources = sourceFiles;
+
+    // default target
+    if (!args.length)
+        args = ["all"];
+
+    auto targets = args
+        .predefinedTargets // preprocess
+        .array;
+
+    if (targets.length > 0)
+    {
+        if (verbose)
+        {
+            log("================================================================================");
+            foreach (key, value; env)
+                log("%s=%s", key, value);
+            log("================================================================================");
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// D build rules
+////////////////////////////////////////////////////////////////////////////////
+
+// TODO: newdelete is probably not needed anymore
+auto newDelete()
+{
+    Rule rule = {
+        target: env["G"].buildPath("newdelete").objName,
+        sources: [env["ROOT"].buildPath("newdelete.c")],
+        command: [env["HOST_CXX"], "-c", "-o$@", "$<"]
+    };
+    return rule;
+}
+
+// Builds the lexer as a separate library
+auto lexer()
+{
+    Rule rule = {
+        target: env["G"].buildPath("lexer").libName,
+        sources: sources.lexer.chain(sources.lexerRoot).array,
+        command: [
+            env["HOST_DMD_RUN"],
+            "-of$@",
+            "-lib",
+            "-J"~env["G"], "-J../res",
+            "-L-lstdc++",
+        ].chain(flags["DFLAGS"], "$<".only).array
+    };
+    return rule;
+}
+
+// Generates a dmd.conf file in the generated folder
+auto dmdConf()
+{
+    // TODO: add support for Windows
+    string exportDynamic;
+    version(OSX) {} else
+        exportDynamic = " -L--export-dynamic";
+
+    auto conf = `[Environment32]
+DFLAGS=-I%@P%/../../../../../druntime/import -I%@P%/../../../../../phobos -L-L%@P%/../../../../../phobos/generated/$(OS)/$(BUILD)/32{exportDynamic}
+
+[Environment64]
+DFLAGS=-I%@P%/../../../../../druntime/import -I%@P%/../../../../../phobos -L-L%@P%/../../../../../phobos/generated/$(OS)/$(BUILD)/64{exportDynamic} -fPIC`.replace("{exportDynamic}", exportDynamic);
+
+    auto target = env["G"].buildPath("dmd.conf");
+    auto commandFun = (){
+        conf.toFile(target);
+    }; // defined separately to support older D compilers
```

Comment:
Due to a recently DMD bug declaring the function directly doesn't work.
See also:

https://issues.dlang.org/show_bug.cgi?id=18199
https://github.com/dlang/dmd/pull/8051